packages/waku/src/router/client.ts

```diff
@@ -606,3 +607,244 @@ class ErrorBoundary extends Component<
+  });
+
+  const routeElement = createElement(Slot, { id: getRouteSlotId(route.path) });
+
+  return createElement(
+    RouterContext.Provider,
+    { value: { route, changeRoute, prefetchRoute } },
+    routeElement,
+  );
+};
+
+export function NewRouter({ routerData = DEFAULT_ROUTER_DATA }) {
+  const route = parseRouteFromLocation();
+  const initialRscPath = encodeRoutePath(route.path);
+  // FIXME cachedIdSetRef and staticPathSetRef should be initialized from the initial RSC payload
+  const cachedIdSetRef = useRef(new Set<string>());
+  const staticPathSetRef = useRef(new Set<string>());
+  const unstable_enhanceCreateData =
+    (
+      createData: (
+        responsePromise: Promise<Response>,
+      ) => Promise<Record<string, ReactNode>>,
+    ) =>
+    async (responsePromise: Promise<Response>) => {
+      const response = await responsePromise;
+      const has404 = routerData[2];
+      if (response.status === 404 && has404) {
+        // HACK this is still an experimental logic. It's very fragile.
+        // FIXME we should cache it if 404.txt is static.
+        return fetchRsc(encodeRoutePath('/404'));
+      }
+      const data = createData(responsePromise);
+      Promise.resolve(data)
+        .then((data) => {
+          if (data && typeof data === 'object') {
+            const {
+              [ROUTE_ID]: routeData,
+              [IS_STATIC_ID]: isStatic,
+              [HAS404_ID]: has404,
+              ...rest
+            } = data;
+            if (routeData) {
+              const [path, query] = routeData as [string, string];
+              // FIXME this check here seems ad-hoc (less readable code)
+              if (
+                window.location.pathname !== path ||
+                window.location.search.replace(/^\?/, '') !== query
+              ) {
+                routerData[1]?.forEach((listener) => listener(path, query));
+              }
+              if (isStatic) {
+                staticPathSetRef.current.add(path);
+              }
+            }
+            if (has404) {
+              routerData[2] = true;
+            }
+            Object.keys(rest).forEach((id) => {
+              cachedIdSetRef.current.add(id);
+            });
+          }
+        })
+        .catch(() => {});
+      return data;
+    };
+  const initialRscParams = JSON.stringify({ query: route.query });
+  return createElement(
+    ErrorBoundary,
+    null,
+    createElement(
+      Root as FunctionComponent<Omit<ComponentProps<typeof Root>, 'children'>>,
+      { initialRscPath, initialRscParams, unstable_enhanceCreateData },
+      createElement(NewInnerRouter, {
+        routerData,
+        cachedIdSetRef,
+        staticPathSetRef,
+      }),
+    ),
+  );
+}
+
+/**
+ * ServerRouter for SSR
+ * This is not a public API.
+ */
+export function NewServerRouter({ route }: { route: RouteProps }) {
+  const routeElement = createElement(Slot, { id: getRouteSlotId(route.path) });
+  return createElement(
+    Fragment,
```

Comment:
What is the purpose of this Fragment?

Reply:
Maybe, it's to match the component tree with NewRouter. If removing it works, we can remove it. TBH, I don't like NewServerRouter hack. I wish just NewRouter worked.